File: downloadablebookandexercisecode/1256/code_files/Chapter_05_Searching_and_Optimization/CSES_FerrisWheel.py
# Ferris Wheel 2
(n, x) = list(map(int, input().split()))
ws = list(map(int, input().split()))

# n - number of kids
# x - maximum allowed weight
# ws - weights
ws.sort()

cnt = 0
i1 = 0
i2 = n - 1

while i1 <= i2:
    # if a smaller kid fits in, we'll put him in as well!
    if ws[i1] + ws[i2] <= x:
        i1 += 1

    # in any case, we are going to sit theheavy kid in a Gondola
    i2 -= 1
    cnt += 1

print(str(cnt))
# Input:
# 4 10
# 7 2 3 9
# Output:
# 3

File: downloadablebookandexercisecode/1256/code_files/Chapter_05_Searching_and_Optimization/CSES_FerrisWheel.cpp
#include <iostream>
#include <algorithm>

using namespace std;

int main() {


    int n, x;
    cin >> n >> x;

    int w[n];
    for (int i = 0; i < n; ++i)
        cin >> w[i];

    sort(w, w + n);

    int i1 = 0;
    int i2 = n - 1;
    int cnt = 0;

    while (i1 <= i2) {
        if (w[i1] + w[i2] <= x)
            i1++;   // We are sitting the lighter person in there as well.
        // at any rate, the heavy weight will sit there for sure
        i2--;
        cnt++;
    }

    cout << cnt;
    return 0;
}
